src/node.rs

```diff
@@ -462,195 +491,225 @@ impl InMemoryNode {
-            .read_value(&balance_key);
-
-        Ok(h256_to_u256(balance)).into_boxed_future()
+    ) -> BoxFuture<Result<U256, jsonrpc_core::Error>> {
+        let inner = Arc::clone(&self.inner);
+
+        Box::pin(async move {
+            let balance_key = storage_key_for_standard_token_balance(
+                AccountTreeId::new(L2_ETH_TOKEN_ADDRESS),
+                &address,
+            );
+
+            match inner.write() {
+                Ok(mut inner_guard) => {
+                    let balance = inner_guard.fork_storage.read_value(&balance_key);
+                    Ok(h256_to_u256(balance))
+                }
+                Err(_) => {
+                    let web3_error = Web3Error::InternalError;
+                    Err(into_jsrpc_error(web3_error))
+                }
+            }
+        })
     }
-
+    /// Returns a block by its number.
+    ///
+    /// # Arguments
+    ///
+    /// * `block_number` - A `BlockNumber` enum variant representing the block number to retrieve.
+    /// * `_full_transactions` - A boolean value indicating whether to retrieve full transactions or not.
+    ///
+    /// # Returns
+    ///
+    /// A `BoxFuture` containing a `jsonrpc_core::Result` that resolves to an `Option` of `zksync_types::api::Block<zksync_types::api::TransactionVariant>`.
     fn get_block_by_number(
         &self,
         block_number: zksync_types::api::BlockNumber,
         _full_transactions: bool,
-    ) -> jsonrpc_core::BoxFuture<
+    ) -> BoxFuture<
         jsonrpc_core::Result<
             Option<zksync_types::api::Block<zksync_types::api::TransactionVariant>>,
         >,
     > {
-        // Currently we support only the 'most recent' block.
-        let reader = self.inner.read().unwrap();
-
-        match block_number {
-            zksync_types::api::BlockNumber::Committed
-            | zksync_types::api::BlockNumber::Finalized
-            | zksync_types::api::BlockNumber::Latest => {}
-            zksync_types::api::BlockNumber::Earliest => {
-                return not_implemented("get_block_by_number__Earliest")
-            }
-            zksync_types::api::BlockNumber::Pending => {
-                return not_implemented("get_block_by_number__Pending")
-            }
-            zksync_types::api::BlockNumber::Number(ask_number) => {
-                if ask_number != U64::from(reader.current_miniblock) {
-                    let not_implemented_format = format!("get_block_by_number__{}", ask_number);
-                    return not_implemented(&not_implemented_format);
+        let inner = Arc::clone(&self.inner);
+
+        Box::pin(async move {
+            let reader = match inner.read() {
+                Ok(r) => r,
+                Err(_) => return Err(into_jsrpc_error(Web3Error::InternalError)),
+            };
+
+            match block_number {
+                zksync_types::api::BlockNumber::Earliest => {
+                    return Err(into_jsrpc_error(Web3Error::NotImplemented));
```

Comment:
Can we keep this logic wrapped inside of `not_implemented(...)` so that we still `println!` the name of the function/arguments that are not implemented? It's been super useful for LambdaClass and other external users with so many APIs not implemented yet.

Reply:
Since `not_implemented` returns a `BoxFuture` I can't use it directly here well. I can change `not_implemented` to not return a future but then all the unimplemented endpoints would change to return `Box::pin(async { Err(not_implemented_error("uninstall_filter")) })` which isn't idiomatic. 

Perhaps you have another approach but since `Web3Error::NotImplemented` is only used here in this file I can just add a `println!("Method {} is not implemented", method_name);` to each of these cases? I don't think any of these are dedicated endpoints rather cases in `get_block_by_number` that need to be implemented. 

```
match block_number {
    zksync_types::api::BlockNumber::Earliest => {
        println!("Method get_block_by_number with BlockNumber::Earliest is not implemented");
        return Err(into_jsrpc_error(Web3Error::NotImplemented));
    }
    zksync_types::api::BlockNumber::Pending => {
        println!("Method get_block_by_number with BlockNumber::Pending is not implemented");
        return Err(into_jsrpc_error(Web3Error::NotImplemented));
    }
    zksync_types::api::BlockNumber::Number(ask_number)
        if ask_number != U64::from(reader.current_miniblock) =>
    {
        println!("Method get_block_by_number with BlockNumber::Number({}) is not implemented", ask_number);
        return Err(into_jsrpc_error(Web3Error::NotImplemented));
    }
    _ => {}
}
```